doc-generator/src/main/scala/Main.scala

```diff
@@ -0,0 +1,135 @@
+    val pb = Process(command = Seq("git", "pull"), cwd = Some(clangExtraDir.toJava))
+    require(pb.! == 0, "git pull failed")
+  }
+
+  def isNotTitle(s: String): Boolean = !s.startsWith("#")
+
+  def fixRepoLinks(s: String): String = {
+    val linkRegex = """\[(.+)\]\((.+\.html)\)""".r("name", "link")
+    linkRegex.replaceAllIn(s, matching => {
+      val name = matching.group("name")
+      val link = matching.group("link")
+      val linkToClangTidy =
+        if (link.matches("https?:\\/\\/.+")) link else s"https://clang.llvm.org/extra/clang-tidy/checks/${link.stripPrefix("/")}"
+      val linkEscaped = linkToClangTidy.replaceAllLiterally(" ", "").trim
+      val nameEscaped = name.trim
+      s"[$nameEscaped]($linkEscaped)"
+    })
+  }
+
+  val patternsWithDocs: Seq[(String, String)] = {
+    val iterator = for {
+      file <- (clangExtraDir / "docs" / "clang-tidy" / "checks").children
+      patternId = file.nameWithoutExtension(includeAll = false)
+      if file.extension.exists(_ == ".rst") && file.nameWithoutExtension != "list"
+      markdownFile = Seq("pandoc", "-t", "commonmark", file.pathAsString).!!
+      content = markdownFile.linesIterator.dropWhile(isNotTitle).mkString(System.lineSeparator())
+      withRepoLinks = fixRepoLinks(content)
+    } yield (patternId, withRepoLinks)
+    iterator.toSeq
+  }
+
+  def categoryFromPatternId(patternId: String): (Pattern.Category, Option[Subcategory]) = patternId match {
+    case s"android-$_" => (Pattern.Category.Security, Some(Pattern.Subcategory.Android))
+    case "cert-err52-cpp" => (Pattern.Category.Security, Some(Pattern.Subcategory.DoS))
+    case s"clang-analyzer-security.insecureAPI$_" =>
+    (Pattern.Category.Security, Some(Pattern.Subcategory.InsecureModulesLibraries))
+    case "clang-analyzer-security" => (Pattern.Category.Security, None)
+    case s"${_}unused$_" => (Pattern.Category.UnusedCode, None)
+    case "misc-unused-alias-decls" => (Pattern.Category.UnusedCode, None)
+    case s"bugprone-$_" => (Pattern.Category.ErrorProne, None)
+    case s"readability-$_" => (Pattern.Category.CodeStyle, None)
+    case s"performance-$_" => (Pattern.Category.Performance, None)
+    case s"portability-$_" => (Pattern.Category.Compatibility, None)
+    case _ => (Pattern.Category.ErrorProne, None)
+  }
+
+  def withoutFamily(patternId: String): String = {
+    Seq(
+      "bugprone",
+      "cppcoreguidelines",
+      "clang-analyzer-cplusplus",
+      "clang-analyzer-core",
+      "clang-analyzer-osx",
+      "hicpp",
+      "misc",
+      "modernize",
+      "objc",
+      "performance",
+      "portability",
+      "readability",
+      "zircon"
+    ).map(s => s"$s-").foldLeft(patternId)((acc, prefix) => acc.stripPrefix(prefix))
+  }
+
+  val patterns = patternsWithDocs.map {
+    case (patternId, _) =>
+      val (category, subcategory) = categoryFromPatternId(patternId)
+      Pattern.Specification(Pattern.Id(patternId), Level.Warn, category, subcategory, None)
```

Comment:
Explain why we use Warn as a category here

Reply:
Done ✅